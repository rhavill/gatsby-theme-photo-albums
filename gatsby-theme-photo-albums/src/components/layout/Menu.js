import React from 'react'
import PropTypes from 'prop-types'
import { useState } from 'react'
import {Link} from 'gatsby'
import {css} from '@emotion/core'
import {withTheme} from 'emotion-theming'

const Menu = ({ links, theme }) => {
  const [isOpen, setIsOpen] = useState(false)
  const handleCheck = () => setIsOpen(!isOpen)
  const listDisplay = isOpen ? 'block' : 'none'

  const ul = css`
    list-style-type: none;
    margin: 0rem;
    padding: 0;
    display: ${listDisplay};
    background-color: ${theme.colors.primary};
  `

  const li = css`
    display: block;
    padding: .5rem;
    /*box-sizing: border-box;*/
    border-top: 1px solid;
  `
  const label = css`
    font-size: 2rem;
    padding: .5rem;
    display: inline-block;
    background-color: ${theme.colors.primary};
  `

  return (
    <nav>
      <label htmlFor='menu-checkbox' css={label}>&#9776;</label>
      <input type='checkbox' id='menu-checkbox' defaultChecked={isOpen} 
        onChange={handleCheck} css={{display: 'none'}} />
      <ul css={ul}>{
        links.map((link, i) => (
          <li key={i} css={li}>
            <Link to={link.path}>
              {link.title}
            </Link>
          </li>))
      }</ul>
    </nav>
  )
}

Menu.propTypes = {
  links: PropTypes.arrayOf(
    PropTypes.shape({
      title: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.node,
      ]).isRequired,
      path: PropTypes.string.isRequired,
    }).isRequired,
  ),
  theme: PropTypes.shape({
    colors: PropTypes.shape({
      primary: PropTypes.string.isRequired
    }).isRequired
  }).isRequired
}

export default withTheme(Menu)