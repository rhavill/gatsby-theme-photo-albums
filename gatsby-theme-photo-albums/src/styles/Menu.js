/** @jsx jsx */
import { jsx } from 'theme-ui'
import PropTypes from 'prop-types'

const Menu = ({children}) => 
  <nav sx={{variant: 'layout.header.menu'}}>
    {children}
  </nav>

Menu.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ]).isRequired,
}

export default Menu