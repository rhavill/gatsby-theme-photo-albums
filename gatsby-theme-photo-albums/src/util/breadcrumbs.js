import compose from 'ramda/src/compose'
import {removeFileExtension, toTitleCase} from './url-text'

const breadcrumbs = (path) => {
  // Set path to empty string for root directory. 
  // Also, remove trailing slash (if it exists) for consistency.
  const parts = path === '/' ? [''] : path.replace(/\/$/, '').split('/')
  return parts
    .map((folder, i) => ({
      // convert to "title case" and remove file extension
      title: folder === '' 
        ? '' 
        : compose(toTitleCase, removeFileExtension)(folder),
      path: folder === '' ? '/' : parts.filter((part, j) => j <= i).join('/') 
    }))
}

export default breadcrumbs