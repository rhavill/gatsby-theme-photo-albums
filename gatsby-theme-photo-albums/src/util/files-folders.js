const curry = require('ramda/src/curry')
const filter = require('ramda/src/filter')
const match = require('ramda/src/match')
const {prependbaseUrl} = require('./url-text')

const getChildPaths = curry(
  (currentPath, paths) => {
    return filter(isParentDirectoryOf(currentPath), paths)
  }
)

/**
 * Determines whether a directory is a parent directory of a file or directory.
 * @param {string} parent The path of a possible parent directory. 
 * @param {string} child The path of a file or directory.
 * @returns {bool}
 */
const isParentDirectoryOf = curry(
  (parent, child) => {
    const regex = new RegExp('^' + parent + '/?[^/]+$')
    return match(regex, child).length > 0
  }
)

const getPagerData = (currentPath, files, filesPerPage) => {
  const pagerData = []
  const children = getChildPaths(currentPath, files)
  const childCount = children.length
  const numPages = Math.ceil(childCount / filesPerPage) || 1
  Array.from({ length: numPages }).forEach((_, i) => {
    pagerData.push({
      limit: filesPerPage,
      skip: i * filesPerPage,
      numPages,
      currentPage: i + 1
    })
  })
  return pagerData
}

/**
 * Accepts an array of gatsby-source-filesystem objects that must have a
 * relativePath property. Adds a url property to all of the objects.
 */
const addUrlProps = curry((baseUrl, filesOrDirectories) => 
  filesOrDirectories.map(fileOrDirectory => ({
    ...fileOrDirectory,
    url: prependbaseUrl(baseUrl, fileOrDirectory.relativePath)
  })))

module.exports = {
  addUrlProps,
  getChildPaths,
  getPagerData,
}