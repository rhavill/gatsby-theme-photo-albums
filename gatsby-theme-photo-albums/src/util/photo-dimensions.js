export default (windowDimensions, pageContext) => {
  const {width, height} = windowDimensions
  const ratio = pageContext.width / pageContext.height
  if (pageContext.width < width && pageContext.height < height)
    return {maxWidth: width, maxHeight: height}
  const excessWidthFactor = (pageContext.width - width) / width
  const excessHeightFactor = (pageContext.height - height) / height
  return excessWidthFactor > excessHeightFactor
    ? {maxWidth: width, maxHeight: width / ratio}
    : {maxWidth: height * ratio, maxHeight: height}
}