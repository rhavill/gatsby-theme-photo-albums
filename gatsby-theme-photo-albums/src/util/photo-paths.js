const forEach = require('ramda/src/forEach')
const keys = require('ramda/src/keys')
const match = require('ramda/src/match')
const {indexedForEach, groupByProp} = require('./ramda-utils')
const {removeFileExtension} = require('./url-text')
const groupByRelativeDirectory = groupByProp('relativeDirectory')

const getPhotoPathsWithPages = (photosPerPage, fileData) => {
  const photoPaths = {}
  const groupedFiles = groupByRelativeDirectory(fileData)
  forEach(relativeDirectory => indexedForEach(
    (file, i) => {
      const pageNumber = Math.ceil((i + 1) / photosPerPage)
      photoPaths[file.url] = getPhotoPathWithPage(
        pageNumber, file.url
      )
    }, 
    groupedFiles[relativeDirectory]
  ), keys(groupedFiles))
  return photoPaths
}

const getPhotoPathWithPage = (pageNumber, url) => {
  url = removeFileExtension(url)
  const matches = match(/^(.*\/)([^/]+)$/, url)
  if (matches && matches[1] && matches[2]) {
    const path =  `${matches[1]}` + (pageNumber > 1 ? `${pageNumber}/` : '') + matches[2]
    return path
  }
  return url
}

module.exports = {getPhotoPathsWithPages, getPhotoPathWithPage}