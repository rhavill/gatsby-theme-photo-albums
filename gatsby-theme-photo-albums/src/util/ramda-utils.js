const addIndex = require('ramda/src/addIndex')
const complement = require('ramda/src/complement')
const compose = require('ramda/src/compose')
const forEach = require('ramda/src/forEach')
const groupBy = require('ramda/src/groupBy')
const isNil = require('ramda/src/isNil')
const map = require('ramda/src/map')
const prop = require('ramda/src/prop')

/**
 * Accepts an array of objects and returns an array containing the values of a 
 * single property from each object of the input array.
 *
 * @func
 * @sig Array -> Array
 * @param {String} property The property to extract from each object of the 
 *    input array.
 * @param {Array} objects The input array containing objects.
 * @return {Array} A simple array containing values exracted from the `property` 
 *    values of the input array.
 * @example
 *
 *      const objects = [{name: 'Hilda', age: 33}, {name: 'José', age: 45}]
 *      objectArrayToPropArray('name', objects); //=> ['Hilda', 'José']
 */
const objectArrayToPropArray = (property, objects) => map(prop(property), objects)

const mapIndexed = addIndex(map)

const indexedForEach = addIndex(forEach)

const groupByProp = property => compose(groupBy, prop)(property)

const isNotNil = complement(isNil)

module.exports = {
  objectArrayToPropArray,
  mapIndexed,
  indexedForEach,
  groupByProp,
  isNotNil,
}
