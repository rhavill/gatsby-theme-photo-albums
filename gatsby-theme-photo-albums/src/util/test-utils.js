import mergeDeepRight from 'ramda/src/mergeDeepRight'
import React from 'react'
import PropTypes from 'prop-types'
import { render } from '@testing-library/react'
import renderer from 'react-test-renderer'
import { ThemeProvider } from 'emotion-theming'

import theme from '../theme'

const create = (element, options) => {
  return renderer.create(
    <ThemeProvider theme={theme}>{element}</ThemeProvider>, options
  )
}

const customRenderer = mergeDeepRight(renderer, {create})

const Wrapper = ({ children }) => {
  return (
    <ThemeProvider theme={theme}>
      {children}
    </ThemeProvider>
  )
}
Wrapper.propTypes = {
  children: PropTypes.node.isRequired
}

const customRender = (ui, options) =>
  render(ui, { wrapper: Wrapper, ...options })

// re-export everything from React Testing Library
export * from '@testing-library/react'

// override react testing library render method and custom react-test-renderer r
export { customRender as render, customRenderer as renderer }