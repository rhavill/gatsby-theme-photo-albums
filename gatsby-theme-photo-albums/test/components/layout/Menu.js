import React from 'react'
import {cleanup, render, renderer} from '../../../src/util/test-utils'
import Menu from '../../../src/components/layout/Menu'

// automatically unmount and cleanup DOM after the test is finished.
afterEach(cleanup)

describe('Menu', () => {
  const links = [
    {title: 'Home', path: '/'},
    {title: 'One', path: '/one'},
    {title: 'Two', path: '/one/two'}
  ]
  it('renders correctly', () => {
    const tree = renderer
      .create(<Menu links={links}  />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
  it('displays "title case" text for path components', () => {
    const {getByText} = render(
      <Menu links={links}  />
    )
    expect(getByText('One')).toBeTruthy()
    expect(getByText('Two')).toBeTruthy()
  })
})
