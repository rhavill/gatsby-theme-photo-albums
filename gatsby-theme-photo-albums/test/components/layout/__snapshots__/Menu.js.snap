// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu renders correctly 1`] = `
<nav>
  <label
    css={
      Object {
        "map": undefined,
        "name": "1ox5w5f",
        "next": undefined,
        "styles": "
    font-size: 2rem;
    padding: .5rem;
    display: inline-block;
    background-color: #6bccf9;
    cursor: pointer;
  ",
        "toString": [Function],
      }
    }
    htmlFor="menu-checkbox"
  >
    â˜°
  </label>
  <input
    css={
      Object {
        "display": "none",
      }
    }
    defaultChecked={false}
    id="menu-checkbox"
    onChange={[Function]}
    type="checkbox"
  />
  <ul
    css={
      Object {
        "map": undefined,
        "name": "1myc36q",
        "next": undefined,
        "styles": "
    list-style-type: none;
    margin: 0rem;
    padding: 0;
    display: none;
    background-color: #6bccf9;
  ",
        "toString": [Function],
      }
    }
  >
    <li
      css={
        Object {
          "map": undefined,
          "name": "lmlfqw",
          "next": undefined,
          "styles": "
    display: block;
    padding: .5rem;
    /*box-sizing: border-box;*/
    border-top: 1px solid;
  ",
          "toString": [Function],
        }
      }
    >
      <a
        href="/"
      >
        Home
      </a>
    </li>
    <li
      css={
        Object {
          "map": undefined,
          "name": "lmlfqw",
          "next": undefined,
          "styles": "
    display: block;
    padding: .5rem;
    /*box-sizing: border-box;*/
    border-top: 1px solid;
  ",
          "toString": [Function],
        }
      }
    >
      <a
        href="/one"
      >
        One
      </a>
    </li>
    <li
      css={
        Object {
          "map": undefined,
          "name": "lmlfqw",
          "next": undefined,
          "styles": "
    display: block;
    padding: .5rem;
    /*box-sizing: border-box;*/
    border-top: 1px solid;
  ",
          "toString": [Function],
        }
      }
    >
      <a
        href="/one/two"
      >
        Two
      </a>
    </li>
  </ul>
</nav>
`;
